xpm/xpm.go

```diff
@@ -43,18 +49,19 @@ func parseXPM(data io.Reader) (image.Image, error) {
 				colors[id] = c
 			}
 		} else {
-			parsePixels(row, charSize, rowNum-colCount-1, colors, img)
+			err := parsePixels(row, charSize, rowNum-colCount-1, colors, img)
+			if err != nil {
+				return nil, err
+			}
 		}
 		rowNum++
 	}
 	return img, scan.Err()
 }
 
 func parseColor(data string, charSize int) (id string, c color.Color, err error) {
-	// TODO: parseColor should return a non-nil err in all error cases,
-	// instead of just returning and implicitly leaving err as nil.
 	if len(data) < charSize {
-		return
+		return "", nil, fmt.Errorf("invalid format: missing color specification")
```

Comment:
You are not using any formatting. Consider using `errors.New()` instead. You might also want to bring it out to a global error variable given that the same error is used in multiple places.

Reply:
I introduced an `ErrInvalidFormat` and return `fmt.Errorf("%w: ...")` in the error cases, so that callers can tell it apart.